packages/rum/src/domain/record/viewports.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Browsers have not standardized various dimension properties. Mobile devices typically report
+ * dimensions in reference to the visual viewport, while desktop uses the layout viewport. For example,
+ * Mobile Chrome will change innerWidth when a pinch zoom takes place, while Chrome Desktop (mac) will not.
+ *
+ * With the new Viewport API, we now calculate and normalize dimension properties to the layout viewport.
+ * If the VisualViewport API is not supported by a browser, it isn't reasonably possible to detect or normalize
+ * which viewport is being measured. Therefore these exported functions will fallback to assuming that the layout
+ * viewport is being measured by the browser
+ */
+
+// Scrollbar widths vary across properties on different devices and browsers
+const TOLERANCE = 10
+
+const isVisualViewportFactoredIn = () =>
+  Math.abs(visualViewport.pageTop - visualViewport.offsetTop - window.scrollY) > TOLERANCE ||
+  Math.abs(visualViewport.pageLeft - visualViewport.offsetLeft - window.scrollX) > TOLERANCE
+
+export const getLayoutViewportDimensions = () => {
+  if (!window.visualViewport || !isVisualViewportFactoredIn()) {
+    const dimensions = {
+      innerWidth: window.innerWidth,
+      innerHeight: window.innerHeight,
+      scrollX: window.scrollX,
+      scrollY: window.scrollY,
+    }
+    // scrollX/Y is subpixel precise, but has less support so needs a fallback
+    if (window.scrollX === undefined || window.scrollY === undefined) {
+      const scrollXFallback =
+        window.pageXOffset !== undefined
+          ? window.pageXOffset
+          : (document.documentElement || document.body.parentNode || document.body).scrollLeft
```

Comment:
**suggestion:** Can you extract variables? It's not super easy to read?
(same for the second one)